components/script/dom/htmlvideoelement.rs

```diff
@@ -177,28 +177,58 @@ impl HTMLVideoElement {
         // network activity as possible.
         let window = self.owner_window();
         let image_cache = window.image_cache();
-        let sender = generate_cache_listener_for_element(self);
-        let cache_result = image_cache.track_image(
+        let cache_result = image_cache.get_cached_image_status(
             poster_url.clone(),
             window.origin().immutable().clone(),
             None,
-            sender,
             UsePlaceholder::No,
         );
 
-        match cache_result {
+        let id = match cache_result {
             ImageCacheResult::Available(ImageOrMetadataAvailable::ImageAvailable {
                 image,
                 url,
                 ..
             }) => {
                 self.process_image_response(ImageResponse::Loaded(image, url), can_gc);
+                return;
             },
+            ImageCacheResult::Available(ImageOrMetadataAvailable::MetadataAvailable(_, id)) => id,
             ImageCacheResult::ReadyForRequest(id) => {
                 self.do_fetch_poster_frame(poster_url, id, can_gc);
+                id
             },
-            _ => (),
-        }
+            ImageCacheResult::LoadError => {
+                self.process_image_response(ImageResponse::None, can_gc);
+                return;
+            },
+            ImageCacheResult::Pending(id) => id,
+        };
+
+        let trusted_node = Trusted::new(self);
+        let generation = self.generation_id();
+        let sender = window.register_image_cache_listener(
+            id,
+            Box::new(move |window: &Window, response| {
+                let trusted_node = trusted_node.clone();
+                window
+                    .as_global_scope()
+                    .task_manager()
+                    .networking_task_source()
+                    .queue(task!(process_image_response: move || {
```

Comment:
Since we're already invoking the callback in the script thread, using a task here just adds latency. We should be able to skip the task and run the actual callback code directly.

Reply:
I gave this a shot (the version of this that failed the merge queue) and it seems that it caused some problems. Looking over the HTML image element code it seems that various parts of the loading algorithm run in tasks already. I think that the code was depending in some way on the order of events being enforced by the task queue. It's not totally clear to me if it's possible to disentangle just these calls from the queue. @jdm, given that the HTML image element was already queuing this response handling as a task, is it okay to maintain this code for now?

I think that the video element loader doesn't present the same problem and I believe I can make the change you requested there.